### ISSUE TITLE
docker modules: fail to deal with docker-py

### ISSUE THREAD

**Issue by baptistedonaux** [ansible/ansible#49395]
##### SUMMARY

`docker_*` Ansible modules uses `docker-py` dependency which requires Python 3. Ansible docker modules can't found `docker-py` dependency which is installed with **pip3** and not **pip**.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
All Docker modules seems affected. Testing with `docker_image` and docker_network`.

##### ANSIBLE VERSION

```
ansible 2.6.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/baptiste/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15+ (default, Oct  2 2018, 22:12:08) [GCC 8.2.0]
```

##### CONFIGURATION
Nothing

##### OS / ENVIRONMENT
Ubuntu 18.10

##### STEPS TO REPRODUCE

```yaml
- apt:
    name: "{{ item }}"
    state: latest
    update_cache: true
  with_items:
    - python3
    - python3-pip

- pip:
    name: docker-py
    executable: pip3
    state: latest

- docker_image:
    name: traefik
```

##### ACTUAL RESULTS

```
FAILED! => {"changed": false, "msg": "Failed to import docker or docker-py - No module named requests.exceptions. Try `pip install docker` or `pip install docker-py` (Python 2.6)"}
```

**Comment by ansibot** [comment]
Hi @baptistedonaux, thank you for submitting this issue!

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/docker](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @akshay196 @danihodovic @dariko @felixfontein @jwitko @kassiansun @tbouvet
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by felixfontein** [comment]
You need to install [docker-py](https://pypi.org/project/docker-py/) or [docker](https://pypi.org/project/docker/) (the latter is preferred) via pip for the same Python version Ansible runs with.

Neither of the two Python libraries requires Python 3. docker-py supports Python 2.6 and newer, while docker supports Python 2.7 and newer.

notabug

**Comment by jaydrogers** [comment]
I am having the same issues with Python 2.7. I am running **Ubuntu 18.04** as my remote and this fails to work for me:

```
- apt:
    state: latest
    name: python-pip

- pip:
    name: docker

- docker_image:
    name: traefik
```

My error is similar, but says `No module named ordered_dict`.
```
FAILED! => {"changed": false, "msg": "Failed to import docker or docker-py - No module named ordered_dict. Try `pip install docker` or `pip install docker-py` (Python 2.6)"}
```

My version of Python on my remote Ubuntu 18.04 server:
```
Python 2.7.15rc1
```

I ran `pip uninstall docker` & `pip uninstall docker-py`. I then took the advice from @felixfontein and used `pip install docker` (rather than `docker-py`) but I am still getting the error above. Is this related?

**Comment by felixfontein** [comment]
@jaydrogers from searching for `No module named ordered_dict` on the web, this seems to be caused by [requests](http://docs.python-requests.org/en/master/), which is used internally by `docker` and `docker-py`. That import has been removed from requests in requests/requests@7fd9267b3bab1d45f5e4ac0953629c5531ecbc55, which apperently made it into version 2.20.0. Which version of `requests` do you have installed?

**Comment by jaydrogers** [comment]
Hi @felixfontein,

Thanks for your help! You pointed me in the right direction that resolved my issue.

For everyone else landing here from a Google search, here is how I resolved this issue:

### 1. I ran `pip list | egrep 'requests'` on my remote machine to see which version I had
It confirmed that I was running a version of [Requests](https://github.com/requests/requests/) that was outdated.
```
requests (2.18.4)
```

### 2. I then uninstalled everything to make sure that it didn't exist using `pip uninstall docker docker-py requests`
Adjust the command above as needed if you don't have these packages already installed.

If you installed things with `apt`, you will need to uninstall using `apt remove python-docker`.

### 3. I confirmed everything was uninstalled using `pip list | egrep 'docker|docker-py|requests'`
This should give you an empty output (or at least a list of packages that do not equal `docker`, `docker-py`, or `requests`).

### 4. I then reinstalled "Python Docker" using `pip` (not `apt`!!!) and also specified the latest version using Ansible
I specified requests to be equal or greater than version`2.20.1` (which is the latest release as of today).

```
- apt:
    state: latest
    name: python-pip

- pip:
    name:
      - docker
      - requests>=2.20.1

- docker_image:
    name: traefik
```

Hopefully this helps! Thanks to @felixfontein for pointing me in the right direction!!

**Comment by felixfontein** [comment]
@jaydrogers I'm glad I could help (and also that this isn't a bug in Ansible ;) )!

**Comment by lifeofguenter** [comment]
I am having the same issue, unfortunately upgrading `requests` did not help.

Ok got it: these steps were required on the target machine, not the local machine :)

**Comment by bradwood** [comment]
Just to be clear, is the docker python module required on the *remote* (ie, target) host, or on the host that is executing the playbook (i.e, my workstation)?

[The docs](https://docs.ansible.com/ansible/2.7/modules/docker_image_module.html) seem to suggest the latter (the local machine): 

> **Requirements**
> The below requirements are needed on **the host that executes this module**.
> 
> python >= 2.6
> docker-py >= 1.7.0
> Please note that the docker-py Python module has been superseded by docker (see here for details). For Python 2.6, docker-py must be used. Otherwise, it is recommended to install the docker Python module. Note that both modules should not be installed at the same time. Also note that when both modules are installed and one of them is uninstalled, the other might no longer function and a reinstall of it is required.
> Docker API >= 1.20

But @jaydrogers suggests that it is the remote machine? Are the doc's wrong?

thanks

**Comment by felixfontein** [comment]
The host that executes the module (not ansible) **is** usually the remote machine (if you don't use `local_action` or `delegate_to: localhost`).

**Comment by bradwood** [comment]
confusing. when I run a playbook on my local workstation, I think of it as running the module locally. Perhaps the working can be make unambiguous. If it confused me, I'll surely confuse someone else.

**Comment by felixfontein** [comment]
Ansible works by transporting code (in form of modules) to a remote machine and executing it there. If you want the general module documentation to be more clear on this, feel free to create an issue (or even better, a PR) againts the [documentation template](https://github.com/ansible/ansible/blob/devel/docs/templates/plugin.rst.j2#L72).

**Comment by jlbribeiro** [comment]
@felixfontein I also stumbled into this issue because I interpreted "on the host that executes this module" as the "host executing the playbook", as @bradwood apparently did (after having tried multiple local `pip install docker`/`docker-py` combos).

Even though the reason for the wording makes perfect sense (given what an "Ansible module" is) I don't think it's _unambiguously_ clear, as a possible (and intuitive for the ones not familiar with Ansible's internals, even if erroneously) interpretation is that the `docker` Python module needs to be installed alongside Ansible so it may be able to communicate with the Docker API.

On a different note, thank you for the amazing work you all do in Ansible. These are the kind of modules that make my life easier every day.

**Comment by felixfontein** [comment]
@jlbribeiro thanks for your feedback! As I've said above, this is not a docker_* problem, but a general Ansible problem. I've created a general docs issue for this: #52977

If you have a concrete idea on how to improve the wording in the template, feel free to comment on the new issue (or even create a PR for it).

**Comment by MartinSotirov** [comment]
I think the problem stems from the fact that the official documentations says that the python docker module should be installed on the host while it's actually required on the target.

**Comment by felixfontein** [comment]
More precisely, the official documentation says "[...] on the host **that executes this module**".

**Comment by deoren** [comment]
I ran into a variation of this issue recently.

For whatever reason I opted to install `docker` using pip's `--user` flag and then had the unfortunate "idea" to use the `-b` or `--become` option. 

This resulted in the "obviously" installed `docker` module being unavailable to the elevated Ansible instance running my playbook. Sharing in case someone has "one of those days" and stumbles across this later on. Hope it helps you as I paid for this reminder with a good bit of "stupid tax", hopefully enough for the both of us. :)
